lib/AST/ASTContext.cpp

```diff
@@ -2709,15 +2715,43 @@ bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
       return false;
   }
 
+  auto validateVersion = [underlyingVersion](const ModuleLoader::ModuleVersionInfo &bestVersionInfo,
+                                             const ModuleLoader::ModuleVersionInfo &versionInfo) {
+    if (!versionInfo.isValid())
+      return false; // The loader didn't attempt to parse a version.
+
+    if (underlyingVersion && !isClangModuleVersion(versionInfo))
+      return false; // We're only matching Clang module versions.
+
+    if (bestVersionInfo.isValid() &&
+        versionInfo.getSourceKind() <= bestVersionInfo.getSourceKind())
+      return false; // This module version's source is lower priority.
+
+    return true;
+  };
+
   if (version.empty()) {
     // If this module has already been successfully imported, it is importable.
     if (getLoadedModule(ModuleName) != nullptr)
       return true;
 
-    // Otherwise, ask whether any module loader can load the module.
+    // Otherwise, ask whether any module loader can load the module,
+    // and record the module version that the succeeding loader
+    // observed.
+    ModuleLoader::ModuleVersionInfo bestVersionInfo;
+    bool moduleFound = false;
```

Comment:
This was a second, unrelated by that I found when implementing the titular change, thanks to our test suite.

Suppose module `Foo` exists in the search paths and specifies user module version `1.0`.
If the first encountered `canImport` query is unversioned:
```
#if canImport(Foo)
...
#endif
```

Followed by a versioned one:
```
#if canImport(Foo, _version: 2.0)
...
#endif
```

The success of the first check will record an unversioned successful `canImport`, which will cause the second check to evaluate to `true`, which is incorrect.

This change causes even unversioned `canImport` checks to track and record the discovered user module version.